// @agentic/triage/dist/triage/connectors.js
/**
 * TriageConnectors - Unified API for Issue/Project/Review Management
 */
import { createBestProvider, createProvider, } from '../providers/index.js';
export class TriageConnectors {
    config;
    _provider = null;
    _initPromise = null;
    issues;
    constructor(config = {}) {
        this.config = config;
        this.issues = new IssueAPI(this);
    }
    async getProvider() {
        if (this._provider)
            return this._provider;
        if (this._initPromise) {
            await this._initPromise;
            return this._provider;
        }
        this._initPromise = this.initializeProvider();
        await this._initPromise;
        return this._provider;
    }
    async initializeProvider() {
        if (this.config.provider === 'jira' && this.config.jira) {
            this._provider = createProvider({
                type: 'jira',
                ...this.config.jira,
            });
        }
        else if (this.config.provider === 'github' && this.config.github) {
            this._provider = createProvider({
                type: 'github',
                repo: this.config.github.repo,
                token: this.config.github.token,
            });
        }
        else {
            this._provider = await createBestProvider({
                repo: this.config.github?.repo,
            });
        }
    }
    async isReady() {
        try {
            const provider = await this.getProvider();
            return provider.isReady();
        }
        catch {
            return false;
        }
    }
}
class IssueAPI {
    connectors;
    constructor(connectors) {
        this.connectors = connectors;
    }
    async create(options) {
        const provider = await this.connectors.getProvider();
        return provider.createIssue(options);
    }
    async get(id) {
        const provider = await this.connectors.getProvider();
        return provider.getIssue(id);
    }
    async update(id, options) {
        const provider = await this.connectors.getProvider();
        return provider.updateIssue(id, options);
    }
    async list(options) {
        const provider = await this.connectors.getProvider();
        return provider.listIssues(options);
    }
    async getStats() {
        const provider = await this.connectors.getProvider();
        return provider.getStats();
    }
}
let _globalConnectors = null;
export function setTriageConnectors(connectors) {
    _globalConnectors = connectors;
}
export function getTriageConnectors() {
    if (!_globalConnectors) {
        _globalConnectors = new TriageConnectors();
    }
    return _globalConnectors;
}
