// @agentic-dev-library/triage@1.1.0/dist/scoring/weights.js
/**
 * Default weights for complexity scoring dimensions
 * These can be calibrated over time based on outcome data
 */
export const DEFAULT_WEIGHTS = {
    files_changed: 0.15,
    lines_changed: 0.1,
    dependency_depth: 0.15,
    test_coverage_need: 0.1,
    cross_module_impact: 0.15,
    semantic_complexity: 0.2,
    context_required: 0.1,
    risk_level: 0.05,
};
export const DEFAULT_THRESHOLDS = {
    trivial: 2.5,
    simple: 5.0,
    moderate: 7.0,
    complex: 8.5,
};
/**
 * Map complexity tier to recommended agent
 */
export function tierToAgent(tier, cursorEnabled = false) {
    switch (tier) {
        case 'trivial':
        case 'simple':
            return 'ollama';
        case 'moderate':
        case 'complex':
            return 'jules';
        case 'expert':
            // Only use cursor if explicitly enabled, otherwise stay with jules
            return cursorEnabled ? 'cursor' : 'jules';
    }
}
/**
 * Determine tier from weighted score
 */
export function scoreToTier(score, thresholds = DEFAULT_THRESHOLDS) {
    if (score <= thresholds.trivial)
        return 'trivial';
    if (score <= thresholds.simple)
        return 'simple';
    if (score <= thresholds.moderate)
        return 'moderate';
    if (score <= thresholds.complex)
        return 'complex';
    return 'expert';
}
/**
 * Calculate weighted score from raw dimension scores
 */
export function calculateWeightedScore(raw, weights = DEFAULT_WEIGHTS) {
    let score = 0;
    for (const [key, weight] of Object.entries(weights)) {
        score += weight * (raw[key] ?? 0);
    }
    return Math.round(score * 100) / 100; // Round to 2 decimal places
}
