// @agentic-dev-library/triage@1.1.0/dist/index.js
/**
 * @agentic-dev-library/triage - AI-Powered Development Automation
 *
 * The triage package provides a comprehensive SDK for automating
 * the entire development lifecycle:
 *
 * - **Issue Triage**: AI-powered assessment, labeling, and planning
 * - **Code Review**: AI-driven PR reviews and feedback handling
 * - **Sprint Planning**: Weighted prioritization, backlog balancing
 *
 * @example SDK Usage
 * ```typescript
 * import { aiGenerate, getTriageTools } from '@agentic-dev-library/triage';
 *
 * // Generate with AI
 * const result = await aiGenerate(prompt, { systemPrompt });
 * ```
 *
 * @packageDocumentation
 * @module @agentic-dev-library/triage
 */
// ============================================================================
// AI SDK Primitives (Schemas, Tools, Handlers)
// ============================================================================
export * as handlers from './handlers/index.js';
export * as schemas from './schemas/index.js';
export * as tools from './tools/index.js';
// ============================================================================
// AI Integration
// ============================================================================
/**
 * AI generation with Vercel AI SDK + Ollama Cloud
 */
export { CLOUD_HOST, DEFAULT_MODEL, generate as aiGenerate, generateWithTools, getModel, getProvider, } from './ai.js';
// ============================================================================
// Model Context Protocol (MCP)
// ============================================================================
export { CONTEXT7_TOOLS, closeMCPClients, createContext7Client, createFilesystemClient, createGitHubClient, createInlineFilesystemClient, createViteReactClient, FILESYSTEM_TOOLS, getAllTools, getContext7Tools, getFilesystemTools, getGitHubTools, getViteReactTools, initializeMCPClients, runAgenticTask, VITE_REACT_TOOLS, } from './mcp.js';
export { createPlaywrightClient, getPlaywrightTools, PLAYWRIGHT_TOOLS } from './playwright.js';
// ============================================================================
// GitHub (MCP)
// ============================================================================
export { addIssueComment, addIssueLabels, areAllChecksPassing, commentOnPR, convertPRToDraft, createCheckRun, createIssueComment, disableAutoMerge, enableAutoMerge, formatAlertsForAI, getCheckRuns, getCodeScanningAlerts, getDependabotAlerts, getIssue, getPRCodeScanningAlerts, getPRReviewComments, getPRReviews, getPullRequest, getRepoContext, markPRReadyForReview, replyToReviewComment, submitPRReview, waitForChecks, } from './octokit.js';
// ============================================================================
// Test Results
// ============================================================================
export { formatForAI, getFailedTests, getLowCoverageFiles, getTestsByFile, getUncoveredFunctions, parseTestReport, } from './test-results.js';
// ============================================================================
// Triage Providers
// ============================================================================
export { BeadsProvider, clearProviders, createBestProvider, createProvider, GitHubProvider, getAllProviders, getCombinedStats, getProvider as getTriageProvider, normalizePriority, normalizeStatus, normalizeType, priorityToNumber, registerProvider, syncAllProviders, } from './providers/index.js';
// ============================================================================
// Triage Unified API & Tools
// ============================================================================
export { addLabelsTool, closeIssueTool, createIssueTool, createTriageConnectors, getCurrentSprintTool, getIssueStatsTool, getIssueTool, getIssueTools, getPRCommentsTool, getProjectTools, getReadyWorkTool, getReviewTools, getTriageTools, ISSUE_TOOL_DEFINITIONS, listIssuesTool, listSprintsTool, PROJECT_TOOL_DEFINITIONS, REVIEW_TOOL_DEFINITIONS, removeLabelsTool, searchIssuesTool, setTriageConnectors, TriageConnectors, updateIssueTool, } from './triage/index.js';
// ============================================================================
// Queue & Scoring (from main)
// ============================================================================
export * from './mcp-server.js';
export * from './queue/index.js';
export * from './scoring/index.js';
export * from './storage/index.js';
// ============================================================================
// Escalation Ladder
// ============================================================================
export * from './escalation/index.js';
