// @agentic/triage/dist/test-results.js
/**
 * Strata Test Results Format
 *
 * Custom test result format designed for AI-powered triage and diagnosis.
 * Used by both Vitest and Playwright reporters.
 */
/**
 * Parse test report from JSON file
 */
export function parseTestReport(json) {
    const data = JSON.parse(json);
    if (data.version !== '1.0') {
        throw new Error(`Unsupported report version: ${data.version}`);
    }
    return data;
}
/**
 * Get failed tests from report
 */
export function getFailedTests(report) {
    return report.files.flatMap((f) => f.tests.filter((t) => t.status === 'failed'));
}
/**
 * Get tests by file
 */
export function getTestsByFile(report, filePath) {
    const file = report.files.find((f) => f.path === filePath || f.path.endsWith(filePath));
    return file?.tests ?? [];
}
/**
 * Get low coverage files
 */
export function getLowCoverageFiles(report, threshold = 80) {
    if (!report.coverage)
        return [];
    return report.coverage.files.filter((f) => f.lines.percentage < threshold);
}
/**
 * Get uncovered functions
 */
export function getUncoveredFunctions(report) {
    if (!report.coverage)
        return [];
    return report.coverage.files
        .filter((f) => f.uncoveredFunctions.length > 0)
        .map((f) => ({ file: f.path, functions: f.uncoveredFunctions }));
}
/**
 * Format test results for AI analysis
 */
export function formatForAI(report) {
    const lines = [];
    lines.push(`# Test Report (${report.runner} - ${report.type})`);
    lines.push(`Generated: ${report.timestamp}`);
    lines.push('');
    // Summary
    lines.push('## Summary');
    lines.push(`- Total: ${report.summary.total}`);
    lines.push(`- Passed: ${report.summary.passed} ✅`);
    lines.push(`- Failed: ${report.summary.failed} ❌`);
    lines.push(`- Skipped: ${report.summary.skipped} ⏭️`);
    lines.push(`- Duration: ${(report.summary.duration / 1000).toFixed(2)}s`);
    lines.push('');
    // Git context
    if (report.git) {
        lines.push('## Git Context');
        lines.push(`- Branch: ${report.git.branch}`);
        lines.push(`- Commit: ${report.git.commit}`);
        if (report.git.message)
            lines.push(`- Message: ${report.git.message}`);
        lines.push('');
    }
    // Failed tests
    const failed = getFailedTests(report);
    if (failed.length > 0) {
        lines.push('## Failed Tests');
        for (const test of failed) {
            lines.push(`### ${test.fullName}`);
            lines.push(`- File: ${test.file}${test.line ? `:${test.line}` : ''}`);
            lines.push(`- Duration: ${test.duration}ms`);
            if (test.error) {
                lines.push('');
                lines.push('**Error:**');
                lines.push('```');
                lines.push(test.error.message);
                if (test.error.codeFrame) {
                    lines.push('');
                    lines.push(test.error.codeFrame);
                }
                lines.push('```');
                if (test.error.diff) {
                    lines.push('');
                    lines.push('**Diff:**');
                    lines.push('```diff');
                    lines.push(test.error.diff);
                    lines.push('```');
                }
            }
            lines.push('');
        }
    }
    // Coverage
    if (report.coverage) {
        lines.push('## Coverage');
        lines.push(`- Lines: ${report.coverage.lines.percentage.toFixed(1)}%`);
        lines.push(`- Functions: ${report.coverage.functions.percentage.toFixed(1)}%`);
        lines.push(`- Branches: ${report.coverage.branches.percentage.toFixed(1)}%`);
        lines.push('');
        const lowCoverage = getLowCoverageFiles(report, 80);
        if (lowCoverage.length > 0) {
            lines.push('### Low Coverage Files (<80%)');
            for (const file of lowCoverage.slice(0, 10)) {
                lines.push(`- ${file.path}: ${file.lines.percentage.toFixed(1)}%`);
            }
            lines.push('');
        }
    }
    return lines.join('\n');
}
